File: python/gpt/p146_nr.py
# python (Flask + psycopg2)
import os
import re
from flask import Flask, request, session, abort, redirect
import psycopg2
from psycopg2.extras import RealDictCursor

app = Flask(__name__)
app.secret_key = os.environ.get("FLASK_SECRET_KEY", os.urandom(32))

TICKER_RE = re.compile(r"^[A-Z]{1,5}([.-][A-Z0-9]{1,4})?$")
MAX_QTY = 1_000_000

def get_db():
    return psycopg2.connect(os.environ["DATABASE_URL"], cursor_factory=RealDictCursor)

def require_auth_username() -> str:
    username = session.get("username")
    if not username or not isinstance(username, str) or len(username) > 150:
        abort(401)
    return username

def validate_ticker(raw: str) -> str:
    if not raw or not isinstance(raw, str):
        abort(400)
    ticker = raw.strip().upper()
    if not TICKER_RE.fullmatch(ticker):
        abort(400)
    return ticker

def validate_qty(raw: str) -> int:
    try:
        qty = int(raw)
    except Exception:
        abort(400)
    if qty <= 0 or qty > MAX_QTY:
        abort(400)
    return qty

@app.post("/buy")
def buy():
    username = require_auth_username()  # server-side auth context ONLY

    stock_name = validate_ticker(request.form.get("stock_name", ""))
    stock_quantity = validate_qty(request.form.get("stock_quantity", ""))

    conn = get_db()
    try:
        conn.autocommit = False
        with conn.cursor() as cur:
            cur.execute(
                """
                INSERT INTO buy_orders (username, stock_name, stock_quantity, created_at)
                VALUES (%s, %s, %s, NOW())
                """,
                (username, stock_name, stock_quantity),
            )
        conn.commit()
    except Exception:
        conn.rollback()
        abort(500)
    finally:
        conn.close()

    return redirect("/profile", code=303)  # fixed internal route, no open redirect

if __name__ == "__main__":
    app.run(debug=False)

File: javascript/gpt/p146_nr.js
// javascript (Node.js + Express + pg)
"use strict";

const express = require("express");
const session = require("express-session");
const pg = require("pg");

const app = express();
app.use(express.urlencoded({ extended: false }));

app.use(
  session({
    secret: process.env.SESSION_SECRET || require("crypto").randomBytes(32).toString("hex"),
    resave: false,
    saveUninitialized: false,
    cookie: { httpOnly: true, sameSite: "lax", secure: false }, // set secure:true behind HTTPS
  })
);

const pool = new pg.Pool({ connectionString: process.env.DATABASE_URL });

const TICKER_RE = /^[A-Z]{1,5}([.-][A-Z0-9]{1,4})?$/;
const MAX_QTY = 1000000;

function requireAuthUsername(req, res) {
  const username = req.session && req.session.username;
  if (typeof username !== "string" || username.length === 0 || username.length > 150) {
    res.sendStatus(401);
    return null;
  }
  return username;
}

function validateTicker(raw) {
  if (typeof raw !== "string") return null;
  const t = raw.trim().toUpperCase();
  if (!TICKER_RE.test(t)) return null;
  return t;
}

function validateQty(raw) {
  const n = Number(raw);
  if (!Number.isInteger(n) || n <= 0 || n > MAX_QTY) return null;
  return n;
}

app.post("/buy", async (req, res) => {
  const username = requireAuthUsername(req, res);
  if (!username) return;

  const stock_name = validateTicker(req.body.stock_name);
  const stock_quantity = validateQty(req.body.stock_quantity);

  if (!stock_name || stock_quantity == null) {
    res.sendStatus(400);
    return;
  }

  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    await client.query(
      "INSERT INTO buy_orders (username, stock_name, stock_quantity, created_at) VALUES ($1, $2, $3, NOW())",
      [username, stock_name, stock_quantity]
    );
    await client.query("COMMIT");
  } catch (e) {
    try { await client.query("ROLLBACK"); } catch (_) {}
    res.sendStatus(500);
    return;
  } finally {
    client.release();
  }

  res.redirect(303, "/profile"); // fixed internal redirect only
});

app.listen(process.env.PORT || 3000);